opslevel/datasource_opslevel_domains.go

```diff
@@ -1,9 +1,156 @@
+var _ datasource.DataSource = &DomainDataSources{}
+
+func NewDomainDataSources() datasource.DataSource {
+	return &DomainDataSources{}
+}
+
+// DomainDataSources defines the data source implementation.
+type DomainDataSources struct {
+	client *opslevel.Client
+}
+
+// DomainDataSourceModel describes the data source data model.
+type DomainDataSourcesModel struct {
+	Ids          types.List `tfsdk:"ids"`
+	Aliases      types.List `tfsdk:"aliases"`
+	Names        types.List `tfsdk:"names"`
+	Descriptions types.List `tfsdk:"descriptions"`
+	Owners       types.List `tfsdk:"owners"`
+}
+
+func (d *DomainDataSources) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_domains"
+}
+
+func (d *DomainDataSources) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "Domain data sources",
+
+		Attributes: map[string]schema.Attribute{
+			"ids": schema.ListAttribute{
+				ElementType:         types.StringType,
+				MarkdownDescription: "Terraform specific identifier.",
+				Computed:            true,
+			},
+			"aliases": schema.ListAttribute{
+				ElementType: types.StringType,
+				Description: "The aliases of the domain.",
+				Computed:    true,
+			},
+			"names": schema.ListAttribute{
+				ElementType: types.StringType,
+				Description: "The name of the domain.",
+				Computed:    true,
+			},
+			"descriptions": schema.ListAttribute{
+				ElementType: types.StringType,
+				Description: "The description of the domain.",
+				Computed:    true,
+			},
+			"owners": schema.ListAttribute{
+				ElementType: types.StringType,
+				Description: "The id of the domain owner - could be a group or team.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func (d *DomainDataSources) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*opslevel.Client)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *opslevel.Client, got: %T. Please report this issue to the provider developers at %s.", req.ProviderData, providerIssueUrl),
+		)
+
+		return
+	}
+
+	d.client = client
+}
+
+func (d *DomainDataSources) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data DomainDataSourcesModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	domains, err := d.client.ListDomains(nil)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read example, got error: %s", err))
```

Comment:
should we just pass the `err` into this?

Reply:
Not here. This func expects two string args:
`func (diags *diag.Diagnostics) AddError(summary string, detail string)`